apicurio-registry/pkg/controller/apicurioregistry/control_loop.go

```diff
@@ -0,0 +1,142 @@
+package apicurioregistry
+
+import (
+	"context"
+	"errors"
+	ar "github.com/apicurio/apicurio-operators/apicurio-registry/pkg/apis/apicur/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ reconcile.Reconciler = &ApicurioRegistryReconciler{}
+
+type ApicurioRegistryReconciler struct {
+	client           client.Client
+	scheme           *runtime.Scheme
+	controlFunctions []ControlFunction
+	notInitialized   bool
+	ctx              *Context
+	controller       controller.Controller
+}
+
+func NewApicurioRegistryReconciler(mgr manager.Manager) *ApicurioRegistryReconciler {
+
+	return &ApicurioRegistryReconciler{
+		client:           mgr.GetClient(),
+		scheme:           mgr.GetScheme(),
+		controlFunctions: []ControlFunction{},
+		notInitialized:   true,
+	}
+}
+
+func (this *ApicurioRegistryReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("ApicurioRegistryReconciler executing.")
+
+	// GetConfig the spec
+	spec := &ar.ApicurioRegistry{}
+	err := this.client.Get(context.TODO(), request.NamespacedName, spec)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Init
+	if this.notInitialized {
+		this.ctx = NewContext(this.controller, this.scheme, reqLogger, this.client)
+
+		var cf ControlFunction
+		cf = NewDeploymentCF(this.ctx)
+		this.AddControlFunction(cf)
+
+		cf = NewServiceCF(this.ctx)
+		this.AddControlFunction(cf)
+
+		cf = NewIngressCF(this.ctx)
+		this.AddControlFunction(cf)
+
+		cf = NewImageConfigCF(this.ctx)
+		this.AddControlFunction(cf)
+
+		cf = NewConfReplicasCF(this.ctx)
+		this.AddControlFunction(cf)
+
+		this.notInitialized = false
+	}
+
+	// Context update
+	this.ctx.update(spec)
+
+	// GetConfig possible config errors
+	if e, s := this.ctx.configuration.GetErrors(); e {
+		err := errors.New(s)
+		log.Error(err, s)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// The LOOP
+	requeue := false
+	for _, v := range this.controlFunctions {
```

Comment:
The control loop is not clear enough for me. 

The reconcilliation pattern you should follow for each resource object the operator manages is (in pseudo code):
```
object = k8sclient.get(objectname)
if !exists(object) {
  k8sclient.create(object)
  return
}

if updateNeeded(object, desiredObject) {
  newObject = applyPatches(object, desiredObject)
  k8sclient.update(newObject)
}
```

For each object type, you may have specific implementation of those abstract methods like `updateNeeded` or `applyPatches`. For instance, the 3scale operator has some objects it does not reconcilliate, it just deploys them initially. So `updateNeeded` method always returns `false`.

One of the design decision might be being very restrictive and desiredObject must match exactly existing object. Another design decision might be only  reconciliate specific attributes of the object, for instance the replica count, which is directly affected by the `ApicurioRegistry` CR fields. This kind of decision are represented in the  implementation of that `updateNeeded` method.